File: js/395.js
/**
 * @param {string} s 
 * @param {number} k 
 * @return {number}
 */
var longestSubstring = function(s, k) {
  const count = (left, right) => {
    if (right - left + 1 < k) return 0
    let times = new Map()
    // 统计各单词出现频率
    for (let i = left; i <= right; i++) {
      if (times.has(str[i])) {
        times.set(str[i], times.get(str[i])+1)
      } else {
        times.set(str[i], 1)
      }
    }
    // 如果该字符出现频次小于K, 则不可能出现在结果子串中
    // 分别排除，挪动两个指针
    while(right - left + 1 >= k && times.get(str[left]) < k) {
      ++left
    }
    while(right - left + 1 >= k && times.get(str[right]) < k) {
      --right
    }
    // 排除子串中包含符合要求的字符， 但是这个长度不会出现对应的概率
    if(right - left + 1 < k) return 0

    // debugger
    // 再对子串进行处理
    for (let i = left; i <= right; i++) {
      // 如果发现不满足的字符出现在子串中，则分别查看改字符左右两边
      if (times.get(str[i] < k)) {
        return Math.max(count(left, i-1), count(i+1, right))
      }
    }
    return right - left + 1
  }
  const str = s.split('')
  return count(0, s.length - 1)
}

File: java/395.java
class Solution {
  public int longestSubstring(String s, int k) {
    int len = s.length();
    if (len == 0 || k > len) return 0;
    if (k < 2) return len;

    return count(s.toCharArray(), k, 0, len - 1)
  }

  private static int count(char[] chars, int k, int p1, int p2) {
    if (p2 - p1 + 1 < k) return 0;
    int[] times = new int[26]; // 26个字母
    // 统计出现频次
    for (int i = p1; i <= p2; ++i) {
      ++times[char[i] - 'a'];
    }
    // 如果该字符出现频次小于k,则不可能出现在结果子串中
    // 分别排除， 然后挪动两个指针
    while(p2 - p1 + 1 >= k && times[char[p1] - 'a'] < k) {
      ++p1;
    }
    while(p2 - p1 + 1 >= k && times[char[p2] - 'a'] < k) {
      --p2;
    }
    if (p2 - p1 + 1 < k) return 0
    // 得到临时子串，在递归处理
    for (int i = p1; i <= p2; ++i) {
      // 如果第i个不符合要求，切分成左右两段分别递归求得
      if (times[char[i] - 'a'] < k) {
        return Math.max(count(chars, k, p1, i-1), count(chars, k, i+1, p2))
      }
    }
    return p2 - p1 + 1;
  }
}